infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -62,159 +78,160 @@ public class SequencerServerCache implements AutoCloseable {
-        return size() * (ENTRY_OVERHEAD + CONFLICTTXSTREAM_OBJ_SIZE);
+    public int size() {
+        return conflictKeyMap.size();
     }
 
-    /*
-     * Put a value in the cache
-     *
-     * @param conflictStream conflict stream
+    /**
+     * Put the provided hashed conflict keys into the cache,
+     * and evict older entries as necessary.
+     * @param conflictKeys The conflict keys to add into the cache.
+     * @param txVersion    The timestamp associated with these conflict keys.
      */
-    public boolean put(ConflictTxStream conflictStream) {
-
-        Long val = conflictKeys.getOrDefault(conflictStream, Address.NON_ADDRESS);
-        if (val > conflictStream.txVersion) {
-            log.error("For key {} the new entry address {} is smaller than the entry " +
-                            "address {} in cache. There is a sequencer regression.",
-                    conflictStream, conflictStream.txVersion, conflictKeys.get(conflictStream));
-            return false;
-        }
+    public void put(@NonNull Set<ConflictTxStream> conflictKeys, long txVersion) {
+        // The provided timestamp should not already be present in the cache.
+        Preconditions.checkState(!versionMap.containsKey(txVersion),
+                "txVersion=%s is already present in SequencerServerCache", txVersion);
 
-        cacheEntries.add(conflictStream);
-        conflictKeys.put(conflictStream, conflictStream.txVersion);
+        // The provided timestamp should be greater than the smallest address in the cache.
+        final long smallestAddress = firstAddress();
+        Preconditions.checkArgument(txVersion > smallestAddress,
+                "txVersion=%s should be larger than the smallest address=%s", txVersion, smallestAddress);
 
-        while (conflictKeys.size() > cacheSize) {
-            invalidateSmallestTxVersion();
-        }
-        return true;
+        // The provided timestamp should be greater than any previous timestamp associated provided conflict
+        // key information. This check is performed separately to avoid partially updating the cache.
+        conflictKeys.parallelStream().forEach(conflictKey -> {
+            final long prevVersion = get(conflictKey);
+            Preconditions.checkState(prevVersion < txVersion,
+                    "sequencer regression: conflictKey=%s prevVersion=%s txVersion=%s",
+                    conflictKey, prevVersion, txVersion);
+        });
+
+        conflictKeys.forEach(conflictKey -> {
+            final Long prevVersion = conflictKeyMap.put(conflictKey, txVersion);
+
+            // Remove the previous entry if conflict key was present with an older version
+            if (prevVersion != null) {
+                final Set<ConflictTxStream> entries = versionMap.get(prevVersion);
+                entries.remove(conflictKey);
+                if (entries.isEmpty()) {
+                    versionMap.remove(prevVersion);
+                }
+            }
+        });
```

Comment:
Is the following correct: removing partial conflict keys from versionMap is just a memory optimization and not needed for correctness because of the key overwrite that happens in `conflictKeyMap.put(conflictKey, txVersion);` ?

Reply:
In other words, can you omit this logic and rely on evict alone. Making evict account of this might make the logic simpler. It's a little convoluted.

Reply:
That was the original issue being fixed. Since the conflictKeyMap maintains the latest conflict key mappings,  not removing them from the versionMap (previously the priorityQueue) caused their sizes to go out of sync. Hence they are removed here, and repopulated below with the more recent timestamp.

Reply:
Relying in evict is not enough. If the cache is below its capacity (as determined by the number of entries in the conflictKeyMap), then nothing will be evicted. For example, this was the case when we repeatedly put the same KVP in the table, the sizes of both data structures just diverged.

Reply:
Yes, I understand. I was implicitly suggesting relying on evict with extra accounting information. See https://github.com/CorfuDB/CorfuDB/pull/3010#discussion_r959016409

Reply:
This section was simplified after reworking and incorporating the use of a LinkedHashMap instead.